Makefile.dep

```diff
@@ -811,3 +811,7 @@ USEPKG := $(sort $(USEPKG))
 ifneq ($(OLD_USEMODULE) $(OLD_USEPKG),$(USEMODULE) $(USEPKG))
   include $(RIOTBASE)/Makefile.dep
 endif
+
+ifneq (,$(filter ecc_%,$(USEMODULE)))
+  USEMODULE += ecc
+endif
```

Comment:
@PeterKietzmann: Is it by intent that this was put out of the recursive dependency tracking and after the sorting/deduplication of `USEMODULE` entries? This looks like a bug to me.

Reply:
I don't quite understand. What would be the fix?

Reply:
Move that code above this:
https://github.com/RIOT-OS/RIOT/blob/01fb5335a28551195c31b6c40a1a5185a46694a2/Makefile.dep#L859-L864

Personally, I would move it up even more, so that it ends up above all of theses lines:

https://github.com/RIOT-OS/RIOT/blob/01fb5335a28551195c31b6c40a1a5185a46694a2/Makefile.dep#L837-L864

Reply:
To explain what the bug is: in
``` Makefile
ifneq ($(OLD_USEMODULE) $(OLD_USEPKG),$(USEMODULE) $(USEPKG)) 
   include $(RIOTBASE)/Makefile.dep 
endif 
```

The file `Makefile.dep` is included again until no changes in the used modules have taken place. E.g. lets assume at the top of the file some additional modules would have been added if module `ecc` was enabled, but that code did not trigger as `ecc` was not part of `USEMODULE` at that time. The recursive include makes sure that every dependency is caught. Moving the code below that will prevent this from happening.

Another issue is that this line

``` Makefile
USEMODULE := $(sort $(USEMODULE)) 
```

make sure that every module is only once in `USEMODULE` (it sorts and removes duplicates). If `USEMODULE` would contain duplicates, compilation would fail. E.g. if `ecc` was already part of `USEMODULE` the insertion of `ecc` in line 816 would result in such a duplicate.

Being honest, I'm not really an expert in the build system. Maybe @kaspar030 or @cladmi could confirm if I got the details right.

Reply:
I was wrong,  duplicates in `USEMODULE` will not always result in build failures. But it will result at least in ugly warnings. E.g. when simply adding `USEMODULE += core_msg` at the bottom of `Makefile.dep` in the current master branch to have it listed twice (`core_msg` is a default module), I get the following output:

```
examples/hello-world $ make
Building application "hello-world" for "nucleo-f303re" with MCU "stm32f3".

"make" -C /home/maribu/Repos/software/RIOT/boards/nucleo-f303re
"make" -C /home/maribu/Repos/software/RIOT/boards/common/nucleo
"make" -C /home/maribu/Repos/software/RIOT/core
/home/maribu/Repos/software/RIOT/Makefile.base:82: target '/home/maribu/Repos/software/RIOT/examples/hello-world/bin/nucleo-f303re/core/msg.o' given more than once in the same rule
/home/maribu/Repos/software/RIOT/Makefile.base:82: target '/home/maribu/Repos/software/RIOT/examples/hello-world/bin/nucleo-f303re/core/msg.o' given more than once in the same rule
```

Reply:
@maribu thanks for your investigation but the problem is still not clear to my. Ergo, I won't open a PR. Maybe you better open it and have the discussion with @kaspar030 or @cladmi there?

Reply:
I'm on it :-)

Reply:
The issue with the dependency being after `OLD_USEMODULE` lines, is that if `ecc` has a dependency, they will not be resolved. So being there is wrong. There is a reason why all the others are before :D 
There are cases with `default implementation handling` that may not work with the current scheme but it is not the case here.

The issue with `core_msg` is that `Makefile.base` does not handle having `USEMODULE` with duplicates when handling `SUBMODULES` and should just be handled there by doing `sort` before listing the `.c` files.